### PR TITLE
feat(endpoints)!: integrate endpoints 2.0

### PR DESCRIPTION
**Pull request by ganeshnj**
## Issue \#

https://github.com/awslabs/aws-sdk-swift/issues/624

## Description of changes

Counterpart: https://github.com/awslabs/smithy-swift/pull/433

### Endpoint struct

Endpoint struct now supports `attributes` and `headers` properties.

```swift
public struct Endpoint {
    let url: String
    let attributes: [String: Any]
    let headers: [String: [String]]
}
```

### EndpointResolver

By default, `EndpointResolver` is a wrapper on top of Common Runtime (CRT) component, but could be overridden by customers using service client configuration that require some variance in how an endpoint is resolved. Both `EndpointResolver` and `EndpointParameter` are service specific and codegen to `AWS<service name>` package.

```swift
public protocol EndpointResolver {
    func resolve(parameters: EndpointParameters) throws -> Endpoint
}

public struct EndpointParameters {
    let region: String?
    let useFIPSEndpoint: Bool
    let useDualStackEndpoint: Bool
    let endpointId: String?
}
```

### Client & Configuration

Service client initializer accepts service specific configuration `<service name>ClientConfigurationProtocol` instead base `AWSClientRuntime.AWSClientConfiguration`. This is breaking change to existing service clients which take a generic client configuration. This change is required to allow service specific configuration parameters and endpoint resolver.

```swift
public class S3Client {
    let config: S3ClientConfigurationProtocol

    public init(config: S3ClientConfigurationProtocol) {
    }
}
```

```swift
public protocol S3ClientConfigurationProtocol : AWSClientRuntime.AWSClientConfiguration {
    var endpointResolver: EndpointResolver { get }
}
```

```swift
public class S3ClientConfiguration: S3ClientConfigurationProtocol {
    // omitted for brevity
    public var endpointResolver: EndpointResolver
    // omitted for brevity
}
```

Client configuration initiazlier now updated with alphatic order of the params which is also a breaking change because previously `region` was always the first parameter and `runtimeConfig` last. This creates confusing and inconsistent experience comparing with other structs we have in the SDK where params follows alphabatic order make it easier to introduce more params without breaking code.

### Middleware

Middleware is going to be service specific to handle customization for endpoint parameters creation. `EndpointMiddleware` is codegen in the service package. Endpoint Parameters are service specific which are also codegen.

For an hypothetical `PutEvents` operation in an example service

#### Request and response

```swift
public struct PutEventsInput: Swift.Equatable {
    let endpointId: String?
}

public struct PutEventsOutputResponse: Swift.Equatable {
}
```

#### EndpointResolverMiddleware

`EndpointResolverMiddleware` is service specific which takes dependency on service specific  `EndpointResolver` & `EndpointParams`. 

```swift
public struct EndpointResolverMiddleware<OperationStackOutput: ClientRuntime.HttpResponseBinding, OperationStackError: ClientRuntime.HttpResponseBinding>: ClientRuntime.Middleware {
    public let id: Swift.String = "EndpointResolverMiddleware"

    let endpointResolver: EndpointResolver

    let endpointParams: EndpointParams

    public init(endpointResolver: EndpointResolver, endpointParams: EndpointParams) {
        self.endpointResolver = endpointResolver
        self.endpointParams = endpointParams
    }

    public func handle<H>(context: Context,
                  input: ClientRuntime.SdkHttpRequestBuilder,
                  next: H) async throws -> ClientRuntime.OperationOutput<OperationStackOutput>
    where H: Handler,
    Self.MInput == H.Input,
    Self.MOutput == H.Output,
    Self.Context == H.Context
    {
        var endpoint: Endpoint
        do {
            endpoint = try endpointResolver.resolve(params: endpointParams)
        } catch {
            throw ClientRuntime.SdkError<OperationStackError>.client(ClientError.unknownError(("Unable to resolve endpoint.")))
        }

        guard let authScheme = endpoint.authScheme(name: "v4") else {
            throw ClientRuntime.SdkError<OperationStackError>.client(ClientError.unknownError(("Unable to resolve endpoint. Missing auth scheme.")))
        }

        let awsEndpoint = AWSEndpoint(endpoint: endpoint, signingName: authScheme["signingName"] as? String, signingRegion: authScheme["signingRegion"] as? String)

        var host = ""
        if let hostOverride = context.getHost() {
            host = hostOverride
        } else {
            host = "\(context.getHostPrefix() ?? "")\(awsEndpoint.endpoint.host)"
        }

        if let protocolType = awsEndpoint.endpoint.protocolType {
            input.withProtocol(protocolType)
        }

        var updatedContext = context
        if let signingRegion = awsEndpoint.signingRegion {
            updatedContext.attributes.set(key: HttpContext.signingRegion, value: signingRegion)
        }
        if let signingName = awsEndpoint.signingName {
            updatedContext.attributes.set(key: HttpContext.signingName, value: signingName)
        }

        input.withMethod(context.getMethod())
            .withHost(host)
            .withPort(awsEndpoint.endpoint.port)
            .withPath(context.getPath())
            .withHeader(name: "Host", value: host)

        return try await next.handle(context: updatedContext, input: input)
    }

    public typealias MInput = ClientRuntime.SdkHttpRequestBuilder
    public typealias MOutput = ClientRuntime.OperationOutput<OperationStackOutput>
    public typealias Context = ClientRuntime.HttpContext
}
```

### EndpointMiddleware usage

Similar to other middleware construction, `EndpointResolverMiddlware` is injected before the `buildStep` with extra codegen `endpointParams` instance that goes to the `init`.

```swift
let endpointParams = EndpointParams(accelerate: config.accelerate ?? false,
    bucket: input.bucket,
    endpoint: config.endpoint,
    forcePathStyle: config.forcePathStyle,
    region: config.region,
    useArnRegion: config.useArnRegion,
    useDualStack: config.useDualStack ?? false,
    useFips: config.useFips ?? false)
operation.buildStep.intercept(position: .before, middleware: EndpointResolverMiddleware<AbortMultipartUploadOutputResponse, AbortMultipartUploadOutputError>(endpointResolver: config.endpointResolver, endpointParams: endpointParams))
```

### Case 1: client init with region

This is simplest way to instantiate a service client by just providing the region as a param.

```swift
let client = try S3Client(region: "us-west-2")
let input = GetObjectInput(bucket: "jangirg-testing", key: "helloworld")
let output = try await client.getObject(input: input)
```

### Case 2: client init with config

Customers can instantiate the config with their endpoint specific configurations such as `useDualStack`, `useFips` or any client level configuration and pass down to the service client init.

```swift
let config = try S3Client.S3ClientConfiguration(region: "us-west-2", useDualStack: true)
let client = S3Client(config: config)
let input = GetObjectInput(bucket: "jangirg-testing", key: "helloworld")
let output = try await client.getObject(input: input)
```

### Case 3: customer endpoint resolver

This is advanced use case where customers can override the `EndpointResolver` with their own implementation. First thing first, `EndpointResolver` protocol must be confirmed by their own implementation. 

```swift
struct CustomEndpointResolver : EndpointResolver {
    func resolve(params: EndpointParams) throws -> Endpoint {
        let properties: [String: AnyHashable] = [
            "authSchemes": [
                ["name": "sigv4",
                 "signingName": "s3",
                 "signingRegion": "us-west-2"]
            ],
        ]
        return try Endpoint(urlString: "http://s3.us-west-2.amazonaws.com", headers: nil, properties: properties)
    }
}
```

Followed by creating the client config by passing confirming struct/class 

```swift
let customResolver = DefaultEndpointResolver()
let config = try S3Client.S3ClientConfiguration(endpointResolver: customResolver, region: "us-west-2")
let client = S3Client(config: config)
let input = GetObjectInput(bucket: "jangirg-testing", key: "helloworld")
let output = try await client.getObject(input: input)
```

## Conventional Commits

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/EndpointResolverGenerator.kt [hunk 1]
```diff
@@ -19,155 +24,46 @@ import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
...
+    private fun renderResolver(writer: SwiftWriter) {
+        writer.addImport(SwiftDependency.CLIENT_RUNTIME.target)
+
+        writer.openBlock("public struct DefaultEndpointResolver: EndpointResolver  {", "}") {
+            writer.write("public init() {}")
+            writer.openBlock(
+                "public func resolve(params: EndpointParams) throws -> \$L {", "}", ClientRuntimeTypes.Core.Endpoint
+            ) {
+                writer.write("fatalError(\"EndpointResolver not implemented\")")
             }
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/EndpointResolverGenerator.kt:66]: **Question**: Coming in a future PR? Or does something else replace this to ensure that a default endpoint resovler is available?

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/EndpointResolverMiddleware.kt [hunk 2]
```diff
@@ -5,34 +5,98 @@
...
+        writer.addImport(SwiftDependency.CLIENT_RUNTIME.packageName)
+        writer.addImport(AWSSwiftDependency.AWS_CLIENT_RUNTIME.packageName)
+        writer.write("let endpoint = try endpointResolver.resolve(params: endpointParams)")
+
+        writer.write("")
+        writer.openBlock("guard let authScheme = endpoint.authScheme(name: \"sigv4\") else {", "}") {
+            writer.write(
+                "throw \$N.client(ClientError.unknownError((\"Unable to resolve endpoint. Missing auth scheme.\")))",
+                errorType
+            )
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/EndpointResolverMiddleware.kt:60]: **Question**: Is the error here always a _missing_ auth scheme? Or could it also be an _unsupported_ scheme?

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/EndpointResolverMiddleware.kt [hunk 3]
```diff
@@ -5,34 +5,98 @@
...
+        writer.write("")
+        writer.write("let awsEndpoint = AWSEndpoint(endpoint: endpoint, signingName: authScheme[\"signingName\"] as? String, signingRegion: authScheme[\"signingRegion\"] as? String)")
 
-    override val name = "EndpointResolverMiddleware"
+        writer.write("")
+        writer.write("var host = \"\"")
+        writer.openBlock("if let hostOverride = context.getHost() {", "} else {") {
+            writer.write("host = hostOverride")
+        }
+        writer.indent().write("host = \"\\(context.getHostPrefix() ?? \"\")\\(awsEndpoint.endpoint.host)\"").dedent()
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/EndpointResolverMiddleware.kt:71]: **Style**: In strings with a lot of escaping, consider using [raw strings](https://kotlinlang.org/docs/strings.html#raw-strings) to simplify:

```
"""host = "\(context.getHostPrefix() ?? "")\(awsEndpoint.endpoint.host)""""
```

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                        // check context params that comes from the input shape
+                        params.add("$memberName: input.${contextParam.memberName.toCamelCase()}")
+                    } else if (param.builtIn.getOrNull() != null) {
+                        // check the built-in params that comes from the client config
+                        val defaultValue = if (param.isRequired) " ?? ${param.defaultValue.getOrNull()}" else ""
+                        params.add("$memberName: config.${param.name.toString().toCamelCase()}$defaultValue")
+                    } else if (param.defaultValue.getOrNull() != null) {
+                        // check if the param has a default value
+                        params.add("$memberName: ${param.defaultValue.getOrNull().toString().toCamelCase()}")
+                    }
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt:68]: **Question**: When might none of these conditions are true? Right now you'd omit that parameter...is that right?

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                }
+        }
+        writer.write("let endpointParams = EndpointParams(${params.joinToString(separator = ", ")})")
+        writer.write("$operationStackName.${middlewareStep.stringValue()}.intercept(position: ${position.stringValue()}, middleware: \$N<\$N, \$N>(${middlewareParamsString()}))", AWSServiceTypes.EndpointResolverMiddleware, output, outputError)
+    }
+
+    private fun middlewareParamsString(): String {
+        return "endpointResolver: config.endpointResolver, endpointParams: endpointParams"
+    }
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt:77]: **Nit**: Should be a `val` instead.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        val updated = transformer.mapShapes(model) { shape ->
+            when (shape) {
+                is ServiceShape -> {
+                    val shapeBuilder = shape.toBuilder()
+                    var trait = shape.getTrait<ClientContextParamsTrait>()
+                    var builder =
+                        trait?.toBuilder() as ClientContextParamsTrait.Builder? ?: ClientContextParamsTrait.builder()
+
+                    shape.getTrait<EndpointRuleSetTrait>()?.ruleSet?.let { ruleSet ->
+                        var endpointRuleSet = EndpointRuleset.fromNode(ruleSet)
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt:37]: **Nit**: Should be `val`s.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import software.amazon.smithy.swift.codegen.integration.SwiftIntegration
+import software.amazon.smithy.swift.codegen.model.getTrait
+
+/**
+ * Transforms the model to add the ClientContextParamsTrait to the service.
+ */
+class AWSClientContextParamsTransformer : SwiftIntegration {
+    override fun preprocessModel(model: Model, settings: SwiftSettings): Model {
+        val transformer = ModelTransformer.create()
+        val updated = transformer.mapShapes(model) { shape ->
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt:28]: **Nit**: Replace `val updated =` with `return`.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt (hunk 4 above)
Comment (ganeshnj) [reply codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt]: Added a clarification.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt (hunk 7 above)
Comment (ganeshnj) [reply codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt]: Fixed.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt (hunk 6 above)
Comment (ganeshnj) [reply codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/model/AWSClientContextParamsTransformer.kt]: fixed.

File: AWSClientRuntime/Sources/Endpoints/AWSEndpoint.swift [hunk 8]
```diff
@@ -24,17 +24,17 @@ public struct AWSEndpoint: Equatable {
...
      This overrides the default signing name used by an SDK client.
      */
-    let signingName: String?
+    public let signingName: String?
     /**
     The region that should be used for signing requests to this endpoint.
     This overrides the default signing region used by an SDK client.
      */
-    let signingRegion: String?
+    public let signingRegion: String?
```
[the remark this answers is not in the document]
Comment (ganeshnj) [reply AWSClientRuntime/Sources/Endpoints/AWSEndpoint.swift:37]: Yes, they have now become convenience properties. Just removing them will result in a bigger refactor given their references all over.

[the remark this answers is not in the document]
Comment (ianbotsf) [reply AWSClientRuntime/Sources/Endpoints/AWSEndpoint.swift:37]: I think I'm missing something. Are these pre-existing fields now backed by the `properties` bag somehow? How will existing code that utilizes them get valid values?

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt [hunk 9]
```diff
@@ -43,8 +43,7 @@ class OperationEndpointResolverMiddleware(
         ctx.service.getTrait<EndpointRuleSetTrait>()?.ruleSet?.let { node ->
             val ruleSet = EndpointRuleset.fromNode(node)
             val staticContextParams = op.getTrait<StaticContextParamsTrait>()?.parameters ?: emptyMap()
-            val clientContextParams =
-                ctx.service.getTrait<ClientContextParamsTrait>()?.parameters ?: emptyMap()
+            ctx.service.getTrait<ClientContextParamsTrait>()?.parameters ?: emptyMap()
```
Comment (ianbotsf) [inline codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt:46]: **Nit**: Looks like this does nothing and can be removed.

**Review by ianbotsf — APPROVED** [review]

File: AWSClientRuntime/Sources/Endpoints/AWSEndpoint.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (ganeshnj) [reply AWSClientRuntime/Sources/Endpoints/AWSEndpoint.swift:37]: Here is what you are looking for in a test case https://github.com/awslabs/aws-sdk-swift/blob/jangirg/feat/endpoints-2/codegen/smithy-aws-swift-codegen/src/test/kotlin/software/amazon/smithy/aws/swift/codegen/EndpointResolverMiddlewareTests.kt#L50-L54

I take back the word `convenience` rather `authScheme` is determined first and then passed to `AWSEndpoint` init.

```swift
guard let authScheme = endpoint.authScheme(name: "sigv4") else {
    throw ClientRuntime.SdkError<OperationStackError>.client(ClientError.unknownError(("Unable to resolve endpoint. Unsupported auth scheme.")))
}

let awsEndpoint = AWSEndpoint(endpoint: endpoint, signingName: authScheme["signingName"] as? String, signingRegion: authScheme["signingRegion"] as? String)
```

**Review by jbelkins — APPROVED** [review]
Reviewed more for training purposes than for correctness or architecture.  Interesting to see how types and their tests are generated.

File: codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt (hunk 9 above)
Comment (ganeshnj) [reply codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt]: fixed.

**Comment by ganeshnj** [comment]
> @ganeshnj so this PR should enable using S3 Transfer Acceleration and fix #610, right?

A new config params will be added that will allow virtual host styling required for S3 Transfer Acceleration.

**Pull request merged**